Support for the "scan" media feature.

mdubet opened the pull request:
#### a74d36b48f3e9b63b0f7384e616cfc97e0fc38d8
<pre>
Support for the &quot;scan&quot; media feature.
<a href="https://bugs.webkit.org/show_bug.cgi?id=114426">https://bugs.webkit.org/show_bug.cgi?id=114426</a>

Reviewed by Aditya Keerthi, Tim Nguyen and Chris Dumez.

Support parsing/matching for the &quot;scan&quot; media feature (Media Queries Level 3).
<a href="https://www.w3.org/TR/mediaqueries-3/#scan">https://www.w3.org/TR/mediaqueries-3/#scan</a>
Like the other implementations, it assumes a modern TV with progressive
display.

* LayoutTests/imported/w3c/web-platform-tests/css/mediaqueries/test_media_queries-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/mediaqueries/test_media_queries-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/mediaqueries/test_media_queries-expected.txt:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/MediaFeatureNames.h:
* Source/WebCore/css/MediaQueryEvaluator.cpp:
(WebCore::scanEvaluate):
* Source/WebCore/css/MediaQueryExpression.cpp:
(WebCore::isValidValueForIdentMediaFeature):
(WebCore::featureWithValidIdent):
(WebCore::isFeatureValidWithoutValue):

Canonical link: <a href="https://commits.webkit.org/252717@main">https://commits.webkit.org/252717@main</a>
</pre>

mdubet commented on Source/WebCore/css/MediaQueryEvaluator.cpp:
```diff
@@ -481,6 +481,33 @@ static bool dynamicRangeEvaluate(CSSValue* value, const CSSToLengthConversionDat
+     * thus the check against "tv".
+     */
+    auto mediaType = frame.view()->mediaType();
+    static constexpr auto tv = ASCIILiteral::fromLiteralUnsafe("tv");
```
The `static` here avoid copying the data onto the stack, but considering that the data is only ~3bytes, I'm not sure it's actually faster.

pxlcoder replied:
I think the check below can just be `equalLettersIgnoringASCIICase(mediaType, "tv"_s)`, similar to `evaluateResolution`.

(cc @cdumez for string best-practices)

cdumez replied:
@pxlcoder is right. Just use "tv"_s to get an ASCIILiteral. If you're using something with "unsafe" in the name, you're probably doing it wrong :)

mdubet replied:
Fixed ✅

pxlcoder commented on Source/WebCore/css/MediaQueryEvaluator.cpp:
```diff
@@ -481,6 +481,33 @@ static bool dynamicRangeEvaluate(CSSValue* value, const CSSToLengthConversionDat
 
+static bool scanEvaluate(CSSValue* value, const CSSToLengthConversionData&, Frame& frame, MediaFeaturePrefix)
+{
+    auto view = frame.view();
```
My understanding is that all ref-counted stack variables in new code should be protected by `Ref` or `RefPtr`.

So this would become `RefPtr view = frame.view()`.

pxlcoder commented on Source/WebCore/css/MediaQueryEvaluator.cpp:
```diff
@@ -481,6 +481,33 @@ static bool dynamicRangeEvaluate(CSSValue* value, const CSSToLengthConversionDat
+     * We are currently supporting Media Queries Level 3,
+     * thus the check against "tv".
+     */
+    auto mediaType = frame.view()->mediaType();
```
This can use the `view` variable from above.

cdumez commented on Source/WebCore/css/MediaQueryEvaluator.cpp:
```diff
@@ -481,6 +481,33 @@ static bool dynamicRangeEvaluate(CSSValue* value, const CSSToLengthConversionDat
+    if (!equalIgnoringASCIICase(mediaType, tv))
+        return false;
+
+    if (!value)
```
You can drop this check here...

cdumez commented on Source/WebCore/css/MediaQueryEvaluator.cpp:
```diff
@@ -481,6 +481,33 @@ static bool dynamicRangeEvaluate(CSSValue* value, const CSSToLengthConversionDat
+    if (!value)
+        return false;
+
+    if (!is<CSSPrimitiveValue>(*value))
```
... and drop the `*` here so that it does the null check for you.

cdumez replied:
But really, I would just write this:
```
auto* primiveValue = dynamicDowncast<CSSPrimitiveValue>(value);
if (!primiveValue)
    return false;
```

No need for earlier `value` null check or for the downcast below.

cdumez commented on Source/WebCore/css/MediaQueryEvaluator.cpp:
```diff
@@ -481,6 +481,33 @@ static bool dynamicRangeEvaluate(CSSValue* value, const CSSToLengthConversionDat
+    if (!is<CSSPrimitiveValue>(*value))
+        return false;
+
+    // All known implementations (Blink, Gecko) assume and match "progressive"
```
WebKit comments need to end with a period, as per WebKit coding style.

nt1m commented on Source/WebCore/css/MediaQueryEvaluator.cpp:
```diff
@@ -481,6 +481,29 @@ static bool dynamicRangeEvaluate(CSSValue* value, const CSSToLengthConversionDat
+     */
+    auto mediaType = view->mediaType();
+    if (!equalLettersIgnoringASCIICase(mediaType, "tv"_s))
+        return false;
```
```suggestion
    if (!equalLettersIgnoringASCIICase(view->mediaType(), "tv"_s))
        return false;
```

nt1m commented on Source/WebCore/css/MediaQueryEvaluator.cpp:
```diff
@@ -481,6 +481,29 @@ static bool dynamicRangeEvaluate(CSSValue* value, const CSSToLengthConversionDat
+
+    // All known implementations (Blink, Gecko) assume and match "progressive".
+    auto& primitiveValue = downcast<CSSPrimitiveValue>(*value);
+    return primitiveValue.valueID() == CSSValueProgressive;
```
Optional but you can use `dynamicDowncast` as Chris mentioned:

```
    auto* primitiveValue = dynamicDowncast<CSSPrimitiveValue>(value);
    if (!primitiveValue)
        return false;

    // All known implementations (Blink, Gecko) assume and match "progressive".
    return primitiveValue->valueID() == CSSValueProgressive;
```

mdubet committed:
Support for the "scan" media feature.
https://bugs.webkit.org/show_bug.cgi?id=114426

Reviewed by Aditya Keerthi, Tim Nguyen and Chris Dumez.

Support parsing/matching for the "scan" media feature (Media Queries Level 3).
https://www.w3.org/TR/mediaqueries-3/#scan
Like the other implementations, it assumes a modern TV with progressive
display.

* LayoutTests/imported/w3c/web-platform-tests/css/mediaqueries/test_media_queries-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/mediaqueries/test_media_queries-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/mediaqueries/test_media_queries-expected.txt:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/MediaFeatureNames.h:
* Source/WebCore/css/MediaQueryEvaluator.cpp:
(WebCore::scanEvaluate):
* Source/WebCore/css/MediaQueryExpression.cpp:
(WebCore::isValidValueForIdentMediaFeature):
(WebCore::featureWithValidIdent):
(WebCore::isFeatureValidWithoutValue):

Canonical link: https://commits.webkit.org/252717@main
